knimin/lib/qiita_jira_util.py

```diff
@@ -367,13 +381,163 @@ def create_sequencing_run(pool_id, email, sequencer, reagent_type,
+        # The files are named with the run_name
+        run_name = prep.run_name.iloc[0]
+        # Start with the forward reads
+        full_fps.extend(
+            [(fp, 'raw_forward_seqs')
+             for fp in glob(path_builder('%s_*_R1_*.fastq.gz' % run_name))])
+        # Add the reverse reads
+        full_fps.extend(
+            [(fp, 'raw_reverse_seqs')
+             for fp in glob(path_builder('%s_*_R2_*.fastq.gz' % run_name))])
+        # Add the index reads
+        full_fps.extend(
+            [(fp, 'raw_barcodes')
+             for fp in glob(path_builder('%s_*_I1_*.fastq.gz' % run_name))])
+    else:
+        # Shotgun prep. The output is per sample fastq
+        atype = 'per_sample_FASTQ'
+        # The samples are prefixed with the study id - take advantage of that
+        # Start with the forward reads
+        full_fps.extend(
+            [(fp, 'raw_forward_seqs')
+             for fp in glob(path_builder('%s_*_R1_*.fastq.gz' % study_id))])
+        # Add the reverse reads
+        full_fps.extend(
+            [(fp, 'raw_reverse_seqs')
+             for fp in glob(path_builder('%s_*_R2_*.fastq.gz' % study_id))])
+
+    qiita_study_path = join(config.qiita_uploads_dir, study_id)
+
+    filepaths = []
+    for full_fp, fp_type in full_fps:
+        file_name = basename(full_fp)
+        copy(full_fp, qiita_study_path)
+        filepaths.append((file_name, fp_type))
+
+    return atype, filepaths
+
+
+def complete_sequencing_run(success, run_id, run_path, logs):
+    """Updates the Jira project and attaches the sequencing data to run_path
+
+    Parameters
+    ----------
+    success : bool
+        Whether the run was successful
+    run_id : int
+        The run id
+    run_path : str
+        Path to the directory with the sequencing files
+    logs : list of str
+        The list of log paths
+    """
+    run = db.read_sequencing_run(run_id)
+
+    # Retrieve the prep template information from the DB)
+    preps = db.generate_prep_information(run_id)
+    studies = [prep.study_id.iloc[0] for prep in preps]
+
+    failures = defaultdict(list)
+    if success:
+        jira_comment = "Sequencing complete. Path to raw files: %s" % run_path
+        # Push the prep information to Qiita
+        for prep in preps:
+            # All the rows have the same value, so I just need to access
+            # to a random row to get the study_id -> 0 is guaranteed to exist
+            study_id = prep.study_id.iloc[0]
+
+            # Copy the sequencing files to relevant Qiita folders
+            atype, filepaths = _copy_sequence_files_to_qiita(prep, run_path)
+
+            dtype = (prep.target_gene.iloc[0]
+                     if 'target_gene' in prep else 'Metagenomics')
+
+            prep, _ = _assess_replicates(prep)
+            prep['created_on'] = prep.created_on.apply(lambda x: x.isoformat())
+            # Cast everything to string: some types generate issues
+            prep = prep.astype(str)
```

Comment:
is this already enforced in the `generate_prep_information` step, and if not, should it be? super nervous about the implicit casts...